server/src/main/java/io/druid/segment/loading/SegmentLoaderLocalCacheManager.java

```diff
@@ -62,7 +63,15 @@ public SegmentLoaderLocalCacheManager(
     this.config = config;
     this.jsonMapper = mapper;
 
-    this.locations = Lists.newArrayList();
+    this.locations = Sets.newTreeSet(new Comparator<StorageLocation>()
+    {
+      @Override
+      public int compare(StorageLocation left, StorageLocation right)
+      {
+        // sorted from empty to full
+        return -Longs.compare(left.available(), right.available());
```

Comment:
It should better be standard `java.lang.Long.compare(right.available(), left.available())`. Sorry that I didn't suggested you this form before.